### DISCUSSION TITLE
Regarding web extensions, 4.0.2, Termux/Android

### DISCUSSION THREAD

**Discussion by darekdeo** [coder/code-server#4816] [General]
Recently I've updated nodejs to latest stable in Termux (16.x) and code-server from 3.x to 4.0.2.

Previously extension manager was connected I think to official visual studio code marketplace:
https://marketplace.visualstudio.com/vscode

After update it is connected to open vsx registry:
https://open-vsx.org/

I think it's not an issue by itself and I bet it's possible to change default extension repository in some configuration?

The problem I have is that most of the extensions are now unavailable, only small amount of base extensions are available for downloading. After update my code-server is detected as "Web" and many extensions aren't marked as compatible with "Web".

Is this a deliberate action? If so, can someone explain what's the purpose of detecting code-server as "Web" instead of "Linux" (even on Termux/Android)? I undestand that providing extensions for "Web" VSC can be useful in case of running VSC in website like github where I as a client don't have acces to what's running on the server (nodejs). But in case of code-server I have full access to installation and nodejs.

I've also tried to install extensions manually (vsix file) with no luck, error printed that code-server couldn't remove directory of previous version of the extension... but directory was missing when I checked. I've created directory manually, tried again then error printed that code-server couldn't remove some hashed-like file of extension, but again nothing was there. Tried it on extension that was previously working on my Termux/Android: https://open-vsx.org/extension/fwcd/kotlin

Allowed "Web" extensions downloaded from code-server marketplace are installing without issues.

Thank you, waiting for info.

**Comment by jsjoeio** [comment]
Our docs have some information about why you can't use VS Code's marketplace [here](https://coder.com/docs/code-server/latest/FAQ#why-cant-code-server-use-microsofts-extension-marketplace). There are docs for using your own extension marketplace via environment variables [here](https://coder.com/docs/code-server/latest/FAQ#how-do-i-use-my-own-extensions-marketplace).

> Is this a deliberate action? If so, can someone explain what's the purpose of detecting code-server as "Web" instead of "Linux" (even on Termux/Android)

The web extension was introduced in VS Code after Codespaces came out. There are docs [here](https://code.visualstudio.com/api/extension-guides/web-extensions).

I'm not sure why your environment (Termux) would only let web extensions work. Both types should work 🤔 Were you able to get other extensions work (non-web)?

**Reply by darekdeo** [reply]
Thanks for info regarding marketplace.

> I'm not sure why your environment (Termux) would only let web extensions work. Both types should work 🤔 Were you able to get other extensions work (non-web)?
> 

That's the main issue, I've been switching couple times between 3.x code-server and 4.0.2.

What I've noticed:

1:
4.0.2 has grayed out button for non Web extensions.

3.x has enabled button for every extension.

2:
The reason why I've updated from 3.x to 4.0.2 in the first place was some of previously working extensions stopped working, like this one: https://marketplace.visualstudio.com/items?itemName=asvetliakov.vscode-neovim

3:
Now can't install extensions at all on 3.x (haven't tried old versions). No error message, button changes label to installing but nothing happens.

Can only install web extensions on 4.0.2. When trying to install extension manually I am getting error like below:
![Screenshot_20220202-222620_Chrome.jpg](https://user-images.githubusercontent.com/9741824/152240234-d330758d-6e73-4740-9cbc-dd33baa1f34e.jpg)

**Reply by jsjoeio** [reply]
The fact that it can't delete an existing folder makes me think it's a permissions issue. Have you tried deleting it manually?

**Reply by jsjoeio** [reply]
I wonder if it's possible to see if this replicates using Codespaces or vscode.dev - what do you think?

**Reply by darekdeo** [reply]
I thought it is permission issue too, but as I wrote in the first post:

1. There was no directory to delete in the first place
2. I've created that directory manually
3. Tried to install extension again
4. Different error did show up, that installer couldnt delete some file (every time file was different)
5. Directory I've created previously has been deleted by installer.

Regarding codespaces or vscode.dev - never tried any of these.

**Reply by jsjoeio** [reply]
Ah, sorry I misread. Hmm 🤔 Can you try re-installing code-server on Termux? And then seeing if the same issue happens.

> Regarding codespaces or vscode.dev - never tried any of these.

Okay no worries. Hard to tell if this is specific to code-server 🤔

**Reply by darekdeo** [reply]
> Ah, sorry I misread. Hmm 🤔 Can you try re-installing code-server on Termux? And then seeing if the same issue happens.
> 
I did. Again as I wrote in previous post... 😀
I've been uninstalling/installing/switching between 4.0.2 and 3.x.

I think I've had different issue on 3.x while trying to install extension manually (something with signature?). Would need to check again and confirm on 3.x if that would help

**Reply by jsjoeio** [reply]
> I did. Again as I wrote in previous post... 😀

Oh gosh...🤦‍♂️ My apologies. It's been a looong day. So sorry!

Oh right! You said that your neovim extension doesn't work on 3.x. Dang! That's a dealbreaker. Though with the newest version of VSCodeVim, it should be a lot faster than the neovim one since it runs as a web extension.

That would help! Unfortunately I don't have a _ton_ of bandwidth to dig into this deeper, but maybe we can at least isolate the problem together through some discussion.

**Reply by darekdeo** [reply]
That's true, I am running on 4.0.2 vscodevim as it works as web extension.

Groovy syntax highlighting surprisingly works fine for kotlin too.

I don't have much issues with whole this situation, I was mostly trying to understand the problem.

If I will find something more about it, I'll let know.

Whole this Termux/Android/codeserver setup is mostly done for my previous projects code browsing. Super useful to connect from PC to phone, open code server on PC browser and voila. 😀

**Reply by jsjoeio** [reply]
> Whole this Termux/Android/codeserver setup is mostly done for my previous projects code browsing. Super useful to connect from PC to phone, open code server on PC browser and voila.

Super cool!!! I think @code-asher does some fancy stuff connecting to his phone from Linux as well. I'm just waiting for the day where we no longer need laptops and we use our phones for everything.
